### PR TITLE
Hybrid data+pipeline parallelism for LoRAs should use "column-major" layout

### PR DESCRIPTION
**Pull request by jukofyork**
This is what I get when I try to train a LoRA using 3 machines, each with 2 GPUs in:

```
Using topology: {ProcessCoord(pipe=0, data=0): 0, ProcessCoord(pipe=0, data=1): 1, ProcessCoord(pipe=0, data=2): 2, ProcessCoord(pipe=1, data=0): 3, ProcessCoord(pipe=1, data=1): 4, ProcessCoord(pipe=1, data=2): 5}
```

which is a big problem because for LoRAs you actually have several orders of magnitude more data getting sent for the hidden states than you do for the gradient reduction step!

The logic of why this makes sense for full fine-tuning is explained in the comment [here](https://deepspeed.readthedocs.io/en/latest/_modules/deepspeed/runtime/pipe/topology.html#ProcessTopology):

```python
class ProcessTopology:
    """ Manages the mapping of n-dimensional Cartesian coordinates to linear
    indices. This mapping is used to map the rank of processes to the grid
    for various forms of parallelism.

    Each axis of the tensor is accessed by its name. The provided ordering
    of the axes defines the layout of the topology. ProcessTopology uses a "row-major"
    layout of the tensor axes, and so axes=['x', 'y'] would map coordinates (x,y) and
    (x,y+1) to adjacent linear indices. If instead axes=['y', 'x'] was used, coordinates
    (x,y) and (x+1,y) would be adjacent.
    """
.
.
.
class PipeDataParallelTopology(ProcessTopology):
    """ A topology specialization for hybrid data and pipeline parallelism.

        Uses data parallelism on the last dimension to encourage gradient
        reductions to use high-bandwidth intra-node links and lower-volume
        pipeline communications to use low-bandwidth inter-node links.
    """

    def __init__(self, num_pp, num_dp):
        super().__init__(axes=['pipe', 'data'], dims=[num_pp, num_dp])
```

So I have created a `CustomPipeDataParallelTopology` which gets used when we aren't doing full fine-tuning:

```python
class CustomPipeDataParallelTopology(ProcessTopology):
    """
    A topology specialisation for hybrid data+pipeline parallelism optimized for LoRA training:
    - Sends high-volume "per token" hidden states over PCIe/NVLink.
    - Sends lower-volume "per step" LoRA gradient reductions over Ethernet/InfiniBand.
    """
    def __init__(self, num_pp, num_dp):
        # Swap the axes and dims to change the rank mapping
        super().__init__(axes=['data', 'pipe'], dims=[num_dp, num_pp])


class CustomPipelineModule(PipelineModule):
    def __init__(self, layers, full_fine_tune, model=None, **kwargs):
        # Assign to list to avoid registering the nn.Module
        self.model = [model]
        # Hybrid data+pipeline parallelism for LoRAs should use "column-major" layout
        if not full_fine_tune:
            world_size = dist.get_world_size()
            num_stages = kwargs.get('num_stages')
            if num_stages > 1 and world_size > 1 and world_size % num_stages == 0:
                num_dp = world_size // num_stages
                kwargs['topology'] = CustomPipeDataParallelTopology(num_pp=num_stages, num_dp=num_dp)
        super().__init__(layers, **kwargs)
```

and now I get this:

```
Using topology: {ProcessCoord(data=0, pipe=0): 0, ProcessCoord(data=0, pipe=1): 1, ProcessCoord(data=1, pipe=0): 2, ProcessCoord(data=1, pipe=1): 3, ProcessCoord(data=2, pipe=0): 4, ProcessCoord(data=2, pipe=1): 5}
```

and even using 10gbit Ethernet this reduces the training time by 30-50%.

For anybody using 1gbit Ethernet or many more nodes (ie: the "all vs all" contention pattern), the old method would likely have been completely unusable for hybrid data+pipeline parallelism.

### CODE REVIEW

**Comment by tdrussell** [comment]
This change looks good. I'm wondering if it should be an explicit option, rather than always applying for LoRA.

For example, suppose you have 2 machine, 2 GPUs each, and pipeline_stages=2. So by default, all communication over ethernet is between the pipeline stages, and with this change, the ethernet communication would switch entirely to gradient reduction.

Now suppose you have a 70b parameter model, and are training a lora with 1GB of lora weights (somewhere in between rank 32 and rank 64, I believe). The gradient reduction is 2GB of data sent total, because each machine has to send the grads to the other. If you were training at 4096 sequence length, you have 4096 * 8192 * 2 = 67MB of data for activations sent for each micro batch. Double that for backward pass, giving 134MB of data per gradient accumulation step. Now at 16 GAS, you have 134MB * 16 = 2.1GB of data sent between pipeline stages, per step. So approximately the same volume as the gradient reduction for the lora weights.

If you trained for longer sequence length, or smaller lora, or more GAS, or had more than 2 machines, the data volume swings more heavily towards activations/grads between pipeline stages. But you could also train a larger lora, on smaller sequence length, and the data volume would be more towards gradient reduction. Assuming I did this estimation right, I'm not sure it's always worth switching to the column-major topology layout for loras. Maybe it should just be an option. But I don't have multiple machines so I have no way of comparing.

**Comment by jukofyork** [comment]
> This change looks good. I'm wondering if it should be an explicit option, rather than always applying for LoRA.
> 
> For example, suppose you have 2 machine, 2 GPUs each, and pipeline_stages=2. So by default, all communication over ethernet is between the pipeline stages, and with this change, the ethernet communication would switch entirely to gradient reduction.
> 
> Now suppose you have a 70b parameter model, and are training a lora with 1GB of lora weights (somewhere in between rank 32 and rank 64, I believe). The gradient reduction is 2GB of data sent total, because each machine has to send the grads to the other. If you were training at 4096 sequence length, you have 4096 * 8192 * 2 = 67MB of data for activations sent for each micro batch. Double that for backward pass, giving 134MB of data per gradient accumulation step. Now at 16 GAS, you have 134MB * 16 = 2.1GB of data sent between pipeline stages, per step. So approximately the same volume as the gradient reduction for the lora weights.
> 
> If you trained for longer sequence length, or smaller lora, or more GAS, or had more than 2 machines, the data volume swings more heavily towards activations/grads between pipeline stages. But you could also train a larger lora, on smaller sequence length, and the data volume would be more towards gradient reduction. Assuming I did this estimation right, I'm not sure it's always worth switching to the column-major topology layout for loras. Maybe it should just be an option. But I don't have multiple machines so I have no way of comparing.

Yeah, I did consider trying to run this sort of calculation to decide if it would be worth it or not, but it gets complicated by lots of factors like `modules_to_save`, different optimisers, etc.

I'll look to change it to an option now, as I agree that it's not so clear cut for all LoRAs configurations and some may still benefit from the default "row-major" setting.

**Comment by jukofyork** [comment]
1. I changed to use `use_column_major_topology` (with default value of false) now.

2. I've also renamed the class `ColumnMajorParallelTopology(ProcessTopology)` to make it a little easier to see what it is for.

3. I'm not 100% sure if there ever would be a sane case where this would be valid:

```python
assert world_size % num_stages == 0, f"world_size ({world_size}) must be divisible by num_stages ({num_stages})"
```

If you hold off on pushing this then I'll be able to test it later tonight.

**Comment by jukofyork** [comment]
Yeah, it seems to be working on my setup:

```
Using topology: {ProcessCoord(data=0, pipe=0): 0, ProcessCoord(data=0, pipe=1): 1, ProcessCoord(data=1, pipe=0): 2, ProcessCoord(data=1, pipe=1): 3, ProcessCoord(data=2, pipe=0): 4, ProcessCoord(data=2, pipe=1): 5}
```

**Comment by jukofyork** [comment]
I'm getting a bad shutdown at the end of training now, so would gold off merging this until I can see if I can find out what is causing it. Everything still saves OK, and it happens after the `"'TRAINING COMPLETE!'"` message is output.

After reading this post:

https://discuss.pytorch.org/t/issue-with-torchrun-multi-node-ddp-training-process-group-not-destroyed-error/214810

I'm trying this at the end now:

```python
    # Synchronise processes and destroy process group
    dist.barrier()
    if dist.is_initialized():
        dist.destroy_process_group()

    if is_main_process():
        print('TRAINING COMPLETE!')
```

**Comment by tdrussell** [comment]
I'm not sure if it behaves differently with multiple nodes. But I get that error every time I ctrl+c the script. It's never caused any problem, but this is on a single machine.

**Comment by jukofyork** [comment]
> I'm not sure if it behaves differently with multiple nodes. But I get that error every time I ctrl+c the script. It's never caused any problem, but this is on a single machine.

Yeah, that error was just a red-herring and nothing to do with this PR (my attempt above just hung on the ` dist.barrier()` call).

I can confirm this does work fine though and I have successfully trained a couple of models using it now (I've given up trying to figure out the close-down crash for now).

I'll reset this back out from being in the draft state, but looking over it again I think a couple of the names might need changing as they don't fit with the rest of the codebase, eg: `use_column_major_topology` as none of the other flags are named "use_", etc.

Feel free to edit any or suggest changes :)

**Comment by jukofyork** [comment]
Might also be worth having a final review as just had to resolve a conflict that didn't really look like it needed resolving on the `self.model = [model]`  line (which wasn't actually moved so not sure why it couldn't auto merge this?).
